File: MaximalSquare.java
/**
 * LeetCode 221 Maximal Square
 */
class Solution {
    public int maximalSquare(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0] == null || matrix[0].length == 0) {
            return 0;
        }
        int max = 0;

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                if (matrix[i][j] == '1') {
                    int res = findMax(matrix, i, j);
                    max = Math.max(res, max);
                }
            }
        }
        return max * max;
    }

    private int findMax(char[][] matrix, int i, int j) {
        int cur = 1;
        boolean contiuneFlag = true;
        while (cur + i < matrix.length && cur + j < matrix[i].length && contiuneFlag) {
            for (int colRange = j; colRange <= cur + j; colRange++) {
                if (matrix[i + cur][colRange] == '0') {
                    contiuneFlag = false;
                    break;
                }
            }

            for (int rowRange = i; rowRange <= cur + i; rowRange++) {
                if (matrix[rowRange][j + cur] == '0') {
                    contiuneFlag = false;
                    break;
                }
            }

            if (contiuneFlag) {
                cur++;
            }

        }
        return cur;
    }
}

File: MaximalSquare.go
func min(a int,b int) int {
	if (a < b) {
		return a
	}
	return b
}

func max(a int,b int) int {
	if (a > b) {
		return a
	}
	return b
}

func maximalSquare(matrix [][]byte) int {
	if matrix == nil || len(matrix) == 0 {
		return 0
	}
	maxSide := 0
	n := len(matrix)
	m := len(matrix[0])
	dp := make([][]int,n + 1)
	for i := 0;i <= n;i++ {
		dp[i] = make([]int,m + 1)
	}
	for i := 1;i <= n;i++ {
		for j := 1;j <= m;j++ {
			if matrix[i - 1][j - 1] == '1' {
				dp[i][j] = min(min(dp[i - 1][j],dp[i - 1][j - 1]),dp[i][j - 1]) + 1
			}
			maxSide = max(maxSide,dp[i][j])
		}
	}
	return maxSide * maxSide
}
